[quant] Implement forward and backward autograd functions for fake quantize

asl3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81438

### Summary:
This PR implements custom autograd functions for forward and backward to be used in APoT fake quantization. The implementation follows this doc about custom autograd functions: https://pytorch.org/tutorials/beginner/examples_autograd/polynomial_custom_function.html

### Test Plan:
Run tests with: `python test/quantization/core/experimental/test_fake_quantize.py`

asl3 committed:
[quant] Implement forward and backward methods for fake quantize

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement forward and backward autograd functions for fake quantize"




[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement forward and backward autograd functions for fake quantize"




[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/experimental/test_fake_quantize.py:
```diff
@@ -51,7 +55,7 @@ def test_forward(self):
 
         # get X_expected by converting fp -> apot -> fp to simulate quantize -> dequantize
-        X_to_apot = quantize_APoT(X, alpha, gamma, quantization_levels, level_indices)
+        X_to_apot, mask = quantize_APoT(X, alpha, gamma, quantization_levels, level_indices)
```
I don't think we should be returning a mask

asl3 replied:
@jerryzh168 do you think it makes more sense to just compute mask in the autograd forward method?

asl3 committed:
Update on "[quant] Implement forward and backward autograd functions for fake quantize"


### Summary:
This PR implements custom autograd functions for forward and backward to be used in APoT fake quantization.

### Test Plan:
Run tests with: `python test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

jerryzh168 commented on test/quantization/core/experimental/test_fake_quantize.py:
```diff
@@ -72,5 +76,17 @@ def test_forward_exception(self):
+        input = torch.randn(20, dtype=torch.double, requires_grad=True)
+
+        observer = APoTObserver(b=4, k=2)
+        observer.forward(input)
```
nit: you can just do `observer(input)`

jerryzh168 commented on torch/ao/quantization/experimental/fake_quantize_helper.py:
```diff
@@ -0,0 +1,28 @@
+from torch import Tensor
+from torch.ao.quantization.experimental.quantizer import quantize_APoT, dequantize_APoT
+
+class fake_quantize_helper(torch.autograd.Function):
```
maybe: fake_quantize_function

jerryzh168 commented on torch/ao/quantization/experimental/quantizer.py:
```diff
@@ -33,10 +33,13 @@ def quantize(self, tensor2quantize: Tensor):
+                                                                                  self.alpha))
+
+        # convert to APoT int representation for dtype
+        tensor2quantize = tensor2quantize.int()
```
can we move this to `float_to_apot`?

asl3 replied:
float_to_apot deals with each element and i need to change the dtype so i think i need it here

jerryzh168 reviewed (approved):
looks good overall, had a few comments inline

asl3 committed:
Update on "[quant] Implement forward and backward autograd functions for fake quantize"


### Summary:
This PR implements custom autograd functions for forward and backward to be used in APoT fake quantization.

### Test Plan:
Run tests with: `python test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g

asl3 commented:
@pytorchbot merge -g

dzdang commented on torch/ao/quantization/experimental/quantizer.py:
```diff
@@ -33,10 +33,13 @@ def quantize(self, tensor2quantize: Tensor):
-                                                                         self.quantization_levels,
-                                                                         self.level_indices,
-                                                                         self.alpha))
+        tensor2quantize = tensor2quantize.detach().apply_(lambda x: float_to_apot(x,
```
have the lambda function issues with jit been resolved?

asl3 replied:
yep! the jit issues were only caused by the helper function that prints the size of the model, but i can use a different function `state_dict` to get the size of the model

dzdang commented on torch/ao/quantization/experimental/fake_quantize_function.py:
```diff
@@ -0,0 +1,28 @@
+        quantized_result = quantize_APoT(x, alpha, gamma, quantization_levels, level_indices)
+
+        # calculate mask tensor
+        mask = torch.clone(x)
```
why mask needs to be initialized with a clone of x?

asl3 replied:
i used a clone because `apply` modifies tensors in-place. i can just apply on x here though, since the new x would be out of scope after the function ends.

dzdang replied:
mask is 0s or 1s. while it's going to get overwritten later, it's not an intuitive initialization. you can just initialize it to 0 or leave it uninitialized

asl3 replied:
yep sounds good!

dzdang commented on torch/ao/quantization/experimental/fake_quantize_helper.py:
```diff
@@ -0,0 +1,28 @@
+import torch
```
what is the purpose of this file? seems to be a duplicate of `fake_quantize_fucntion.py`?

asl3 replied:
yep removing this

asl3 committed:
Update on "[quant] Implement forward and backward autograd functions for fake quantize"


### Summary:
This PR implements custom autograd functions for forward and backward to be used in APoT fake quantization. The implementation follows this doc about custom autograd functions: https://pytorch.org/tutorials/beginner/examples_autograd/polynomial_custom_function.html

### Test Plan:
Run tests with: `python test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement forward and backward autograd functions for fake quantize"


### Summary:
This PR implements custom autograd functions for forward and backward to be used in APoT fake quantization. The implementation follows this doc about custom autograd functions: https://pytorch.org/tutorials/beginner/examples_autograd/polynomial_custom_function.html

### Test Plan:
Run tests with: `python test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 committed:
Update on "[quant] Implement forward and backward autograd functions for fake quantize"


### Summary:
This PR implements custom autograd functions for forward and backward to be used in APoT fake quantization. The implementation follows this doc about custom autograd functions: https://pytorch.org/tutorials/beginner/examples_autograd/polynomial_custom_function.html

### Test Plan:
Run tests with: `python test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g

asl3 committed:
Update on "[quant] Implement forward and backward autograd functions for fake quantize"


### Summary:
This PR implements custom autograd functions for forward and backward to be used in APoT fake quantization. The implementation follows this doc about custom autograd functions: https://pytorch.org/tutorials/beginner/examples_autograd/polynomial_custom_function.html

### Test Plan:
Run tests with: `python test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g

asl3 committed:
Update on "[quant] Implement forward and backward autograd functions for fake quantize"


### Summary:
This PR implements custom autograd functions for forward and backward to be used in APoT fake quantization. The implementation follows this doc about custom autograd functions: https://pytorch.org/tutorials/beginner/examples_autograd/polynomial_custom_function.html

### Test Plan:
Run tests with: `python test/quantization/core/experimental/test_fake_quantize.py`

[ghstack-poisoned]

asl3 commented:
@pytorchbot merge -g
